network/health_check.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2023 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"knative.dev/pkg/logging"
+)
+
+// ServeHealthProbes sets up liveness and readiness probes.
+// If user sets no probes explicitly via the context then defaults are added.
+func ServeHealthProbes(ctx context.Context) {
+	logger := logging.FromContext(ctx)
+	port := os.Getenv("KNATIVE_HEALTH_PROBES_PORT")
+	if port == "" {
+		port = "8080"
+	}
+
+	server := http.Server{ReadHeaderTimeout: time.Minute, Handler: muxWithHandles(ctx), Addr: ":" + port}
+
+	go func() {
+		go func() {
+			<-ctx.Done()
+			_ = server.Shutdown(ctx)
+		}()
+
+		// start the web server on port and accept requests
+		logger.Infof("Probes server listening on port %s", port)
+
+		if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			logger.Fatal(err)
+		}
+	}()
+}
+
+func muxWithHandles(ctx context.Context) *http.ServeMux {
+	var handler HealthCheckHandler
+	if h := getOverrideHealthCheckHandler(ctx); h != nil {
+		handler = *h
+	} else {
+		defaultHandle := newDefaultProbesHandle(ctx)
+		handler = HealthCheckHandler{
+			ReadinessProbeRequestHandle: defaultHandle,
+			LivenessProbeRequestHandle:  defaultHandle,
+		}
+	}
+	mux := http.NewServeMux()
+	if UserReadinessProbe(ctx) {
+		mux.HandleFunc("/readiness", handler.ReadinessProbeRequestHandle)
+	}
+	if UserLivenessProbe(ctx) {
+		mux.HandleFunc("/health", handler.LivenessProbeRequestHandle)
+	}
+	// Set both probes if user does not want to explicitly set only one.
+	if !UserReadinessProbe(ctx) && !UserLivenessProbe(ctx) {
+		mux.HandleFunc("/readiness", handler.ReadinessProbeRequestHandle)
+		mux.HandleFunc("/health", handler.LivenessProbeRequestHandle)
+	}
+	return mux
+}
+
+// HealthCheckHandler allows to set up a handler for probes. User can override the default handler
+// with a custom one.
+type HealthCheckHandler struct {
+	ReadinessProbeRequestHandle http.HandlerFunc
+	LivenessProbeRequestHandle  http.HandlerFunc
+}
+
+func newDefaultProbesHandle(sigCtx context.Context) http.HandlerFunc {
+	logger := logging.FromContext(sigCtx)
+	once := sync.Once{}
```

Comment:
Why do we need this `once`?

Reply:
Activator has the same approach too for its healthcheck. So since the probe handler could be part of a handler chain (we might want to make it generic) we want the msg to be printed once. Of course I can simplify it for now if we expect SIGTERM is managed only by one handler once.